File: backend-java/src/main/java/com/example/faceattendance/service/StudentService.java
package com.example.faceattendance.service;

import com.example.faceattendance.dto.student.CreateStudentRequest;
import com.example.faceattendance.dto.student.FilterOptionsResponse;
import com.example.faceattendance.dto.student.StudentResponse;
import com.example.faceattendance.dto.student.UpdateStudentRequest;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

/**
 * Student management service contract.
 *
 * Students are profile records only and are not website users.
 */
public interface StudentService {

    /**
     * Creates a new student.
     * Students are stored only in the students table.
     *
     * @param request Student creation request
     * @return Created student response
     */
    StudentResponse create(CreateStudentRequest request);

    /**
     * Returns all students with optional search and
     * course/batch/semester/year/teacher filters (server-side, paginated).
     */
    Page<StudentResponse> getAll(
            Pageable pageable,
            String search,
            String course,
            String batch,
            String semester,
            String year,
            Long teacherId
    );

    /**
     * Distinct course/batch/semester/year values present in the
     * students table — used to build filter dropdowns without
     * hardcoding anything.
     */
    FilterOptionsResponse getFilterOptions();

    /**
     * Returns a student by ID.
     */
    StudentResponse getById(Long id);

    /**
     * Updates a student.
     */
    StudentResponse update(
            Long id,
            UpdateStudentRequest request
    );

    /**
     * Deletes a student.
     */
    void delete(Long id);
}

File: frontend-UI/src/services/studentService.js
import api from './api';

const studentService = {
  /**
   * Paginated list with server-side search + course/batch/semester/year
   * filters — the browser never needs to load the whole table.
   */
  getAll: ({ page = 0, size = 10, search = '', course, batch, semester, year, teacherId } = {}) => {
    const params = { page, size };
    if (search) params.search = search;
    if (course) params.course = course;
    if (batch) params.batch = batch;
    if (semester) params.semester = semester;
    if (year) params.year = year;
    if (teacherId) params.teacherId = teacherId;
    return api.get('/api/students', { params }).then((r) => r.data);
  },

  /** Distinct course/batch/semester/year values for filter dropdowns. */
  getFilterOptions: () =>
    api.get('/api/students/filter-options').then((r) => r.data),

  getById: (id) => api.get(`/api/students/${id}`).then((r) => r.data),

  create: (data) => api.post('/api/students', data).then((r) => r.data),

  update: (id, data) => api.put(`/api/students/${id}`, data).then((r) => r.data),

  remove: (id) => api.delete(`/api/students/${id}`),
};

export default studentService;
